system/include/SSE/smmintrin.h

```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright 2020 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+#ifndef __emscripten_smmintrin_h__
+#define __emscripten_smmintrin_h__
+
+#ifndef __SSE4_1__
+#error "SSE4.1 instruction set not enabled"
+#endif
+
+#include <tmmintrin.h>
+#include <math.h> // For rint and rintf
+
+#define _mm_blend_epi16(__a, __b, __imm8) __extension__ ({     \
+  (__m128i)__builtin_shufflevector((__i16x8)(__m128i)(__a),    \
+                                   (__i16x8)(__m128i)(__b),    \
+                                   (((__imm8) & 1) ?  8 : 0),  \
+                                   (((__imm8) & 2) ?  9 : 1),  \
+                                   (((__imm8) & 4) ? 10 : 2),  \
+                                   (((__imm8) & 8) ? 11 : 3),  \
+                                   (((__imm8) & 16) ? 12 : 4), \
+                                   (((__imm8) & 32) ? 13 : 5), \
+                                   (((__imm8) & 64) ? 14 : 6), \
+                                   (((__imm8) & 128) ? 15 : 7)); })
+
+#define _mm_blend_pd(__a, __b, __imm8) __extension__ ({         \
+  (__m128d)__builtin_shufflevector((__f64x2)(__m128d)(__a),     \
+                                   (__f64x2)(__m128d)(__b),     \
+                                   (((__imm8) & 0x01) ? 2 : 0), \
+                                   (((__imm8) & 0x02) ? 3 : 1)); })
+
+#define _mm_blend_ps(__a, __b, __imm8) __extension__ ({ \
+  (__m128)__builtin_shufflevector((__f32x4)(__m128)(__a), (__f32x4)(__m128)(__b), \
```

Comment:
Why the double casts here?

Reply:
Because those are implemented with #defines, it is to ensure that the parameters implicitly cast through the type, e.g. if one had a `class MyVec { operator __m128(); };` in it) It is basically to emulate as if these were function calls. Although granted the headers might not be exactly strictly diligent about this in all places.

Reply:
Oh interesting, thanks!